Feat/create update endpoint

Evandro-Ladislau opened the pull request:
feat: create update endpoint.

Commit:
create update endpoint

Evandro-Ladislau committed:
:construction: Create update endpoint

Evandro-Ladislau committed:
:construction: Create update endpoint

Evandro-Ladislau committed:
:construction: Create update endpoint

didibrabosa reviewed (commented):
ficou mídia!

didibrabosa commented on route/product_router.py:
```diff
@@ -22,4 +22,10 @@ async def create_product(product: Product):
+    return product_created
+
+@router.put("/products/{product_id}")
+async def update_product(product_id: str, product: Product):
```
async?? pq?

Evandro-Ladislau replied:
Muito basicamente, o async é usado para criar funções que esperam por algo que demora, sem bloquear outros processamentos do sistema.
Porém, como o Arthur mencionou na review, tem coisas mais complexas por trás disso. Como ainda não tenho esse conhecimento mais aprofundado sobre o async, segui a instrução dele e fiz a remoção.

Bem observado Diego, obrigado!

AurelioNM commented on storage/product_storage.py:
```diff
@@ -60,3 +60,29 @@ def save_product(self, product: Product) -> Product:
+            raise 
+        finally:
+            self.db.commit()
+            cursor.close()
```
como vc já esta usando o `with self.db.cursor()`, não precisa colocar o `cursor.close()` aqui. Pq o with já é responsável por fechar o recurso, após o uso
So precisa do `self.db.commit()`

michaelrodriguess replied:
creio que não é preciso usar o cursor.close() aqui por conta que o with self.db.cursor fecha automaticamente

michaelrodriguess commented on storage/product_storage.py:
```diff
@@ -60,3 +60,29 @@ def save_product(self, product: Product) -> Product:
+                        active = {product.active},
+                        created_at = '{product.created_at}',
+                        updated_at = NOW()
+                    WHERE id = '{id}';
```
aqui temos uma possivel falha de SQL injection, por conta da interpolação para os campos do banco de dados. se eu passar um id -> id = '1'; DROP TABLE products; --'
a Query resultante vai ser: 
 'UPDATE products 
 SET 
    name = 'value',
    description = 'value',
    price = 100,
    quantity = 10,
    active = true,
    created_at = '2024-01-01',
    updated_at = NOW()
WHERE id = '1'; DROP TABLE products; --';

E Isso poderia apagar toda tabela products. O ideal é usar a **parametrização** que é  o operador **%s** para evitar o SQL Injection.   Ficaria algo como:
"""UPDATE products SET name = %s WHERE id = %s;""",(product.name, id). Fazendo assim os valores são passados como parâmetros. O cursor.execute com parâmetros usa queries preparadas, e o banco de dados sanitiza automaticamente os valores antes de inseri-los na query.

Evandro-Ladislau replied:
Você tem razão, eu sei disso. Inclusive, comentamos sobre isso no code review. Faltou um pouco de atenção da minha parte mesmo. Obrigado por pontuar!

Evandro-Ladislau committed:
:recycle: - Alterar o  para definir o campo updated_at como datetime.now.
- Ajustar o SQL para receber updated_at como os demais campos.
- Remover o ID do path e utilizá-lo apenas no body.

AurelioNM commented on .env.example:
Pq deletou o env.example? É importante que ele esteja ai pra que quem for executar o projeto saiba quais variaveis precisam ser preenchidas

AurelioNM commented on route/product_router.py:
```diff
@@ -22,4 +22,10 @@ async def create_product(product: Product):
+@router.put("/products", response_model=Product)
+async def update_product(product: Product):
+    logger.info(f"Started UpdateProduct with body={product.model_dump()}")
+    product_updated = service.update_product(product)
```
Importante logar quando a requisição termina, e o retorno dela. Assim como as outras rotas

AurelioNM commented on storage/product_storage.py:
```diff
@@ -60,3 +61,27 @@ def save_product(self, product: Product) -> Product:
+            self.logger.error(f"Failed on update operation. Error: {ex}")
+            raise 
+        finally:
+            self.db.commit()
```
O `self.db.commit()` pode ficar dentro do try, e esse finally pode ser removido.
So adicionar um `self.db.rollback()` no except, pra fechar a transação no DB em caso de erro

Arthuralentejo commented on service/product_service.py:
```diff
@@ -14,4 +15,9 @@ def get_product_by_id(self, id: str) -> Product:
+    
+    def update_product(self, product: Product) -> Product:
+        self.logger.info(f"Updating product with ID {id}...")
+        product.updated_at = datetime.now()
```
Será que é aqui que deveria fazer isso?

AurelioNM replied:
Fui eu que sugeri :)

Pra manter a consistencia do que vai ser setado na coluna de updated_at
E a response da API

Evandro-Ladislau committed:
:recycle: -Renomear arquivo env para env.example
-Adicionar log quando a requisição de update terminar
-Colocar self.db.commit() dentro do try
-Adicionar self.db.rollback() no except
-Remover finally

Evandro-Ladislau committed:
:recycle: -Renomear arquivo env para env.example

Evandro-Ladislau committed:
:recycle: -- Alterado método para retornar o produto atualizado com dados corretos, utilizando get_product_by_id.

Arthuralentejo commented on storage/product_storage.py:
```diff
@@ -60,3 +61,27 @@ def save_product(self, product: Product) -> Product:
+                    WHERE id = %s;
+                    """, (product.name, product.description, product.price, product.quantity, product.active, product.updated_at, product.id))
+            self.db.commit()          
+            return self.get_product_by_id(product.id)
```
Mano aqui vc ta fazendo sua aplicação fazer uma nova query pra buscar o produto que vc atualizou, então toda requisição de update na real acabaria fazendo duas queries,. Agora pensa q seu sistema recebe 10k requests por minuto, isso não parece ser escalável né? 

Acho que vc pode retornar a model que vc ta usando pra salvar, ja que vc ta atualizando a data do updated at dela na service

AurelioNM replied:
Tem um comando SQL que da pra colocar no Update, q é o RETURNING
Ele retorna os valores atualizados das colunas

Evandro-Ladislau committed:
Merge branch 'develop' into feat/create-update-endpoint

Evandro-Ladislau committed:
:recycle: -Retorno dos dados de product atualizados usando RETURNING

Raph2ll commented on route/product_router.py:
```diff
@@ -32,4 +32,12 @@ def create_product(product: Product):
+    return product_created
+
+@router.put("/products", response_model=Product)
+async def update_product(product: Product):
```
Esse async é necessário aq ?

Raph2ll commented on storage/product_storage.py:
```diff
@@ -69,6 +70,29 @@ def save_product(self, product: Product) -> Product:
+                        quantity = %s,
+                        active = %s,
+                        updated_at = %s
+                    WHERE id = %s;
```
Falta o and active = true

AurelioNM replied:
Nao acho q seja obrigatório. O put tem q ser capaz de atualizar qualquer produto, independente se ta ativo ou nao

Evandro-Ladislau committed:
:recycle: -Removendo Async do método

GS0606 commented on route/product_router.py:
sua router tá criando a função com async !

Evandro-Ladislau committed:
Merge branch 'develop' into feat/create-update-endpoint

Evandro-Ladislau committed:
Merge branch 'feat/create-update-endpoint' of https://github.com/ArianeBeltrao/product-service into feat/create-update-endpoint

Evandro-Ladislau committed:
:recycle: - Retornando produto com método de mapeamento e validação de status ativo na atualização de product.

Commit:
:recycle: - Removendo validação de produto ativo na query de atualização
